### PR TITLE
"project finished (I think)."

### PR DESCRIPTION
**Pull request by harris-andy**

### CODE REVIEW

File: HabitTracker.harris-andy/README.md [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+## Project Description
+  - It's a small console CRUD app in which the user can track hobbies which are stored in a local SQLite database.
+  - Built with C#/.Net 8
+
+
+## Usage
+  - Follow the instructions and away you go
+  - i.e. Select from the menu to perform operations such as: viewing all records, inserting, updating and deleting records.
+
+    <img src="./images/gamemenu.png" alt="Game Menu" width="500"/>
```
Comment (chrisjamiecarter) [inline HabitTracker.harris-andy/README.md:16]: If you are just displaying an image you can do this in native markdown syntax:

`Game Menu`

If the width being 500 is important to you, you can use an image editor to resize to 500px 😄 

This way you do not sully your markdown file with html.
Note: I break this rule myself as when I create my README's I like the app name and logo front and center.
[example](https://github.com/chrisjamiecarter/budget)

File: HabitTracker.harris-andy/Program.cs [hunk 2]
```diff
@@ -0,0 +1,592 @@
...
+        }
+        connection.Close();
+    }
+
+    Console.WriteLine("\n\n");
+    Console.WriteLine("Whelp. It's all gone. Now what...?");
+    Thread.Sleep(2000);
+    MainMenu();
+}
+class HobbyRecord
```
Comment (chrisjamiecarter) [inline HabitTracker.harris-andy/Program.cs:585]: I know the project allows for this. But I would really love to see this class in its own `HobbyRecord.cs` file outside of `Program.cs`

File: HabitTracker.harris-andy/Program.cs [hunk 3]
```diff
@@ -0,0 +1,592 @@
+﻿using System.Data;
+using System.Globalization;
+using Microsoft.Data.Sqlite;
+
+string connectionString = @"Data Source=HabitTracker.db";
+
+using (SqliteConnection connection = new SqliteConnection(connectionString))
```
Comment (chrisjamiecarter) [inline HabitTracker.harris-andy/Program.cs:7]: Remember that you can use the simpler "Using Declaration" syntax to save an indentation level against the "Using Statement" syntax:

(Note I only put this is a Microsoft push it in their style guide which I adhere to (try), but it is user preference)

❌ 
```
// Using Statement
using (var connection = new SqliteConnection(connectionString))
{
    // ...
}
```

✔️ 
```
// Using Declaration
using var connection = new SqliteConnection(connectionString);
// ...
```

File: HabitTracker.harris-andy/Program.cs (hunk 2 above)
Comment (chrisjamiecarter) [inline HabitTracker.harris-andy/Program.cs:585]: But, if you are going to have it like this. It needs that blank line between methods/classes.

File: HabitTracker.harris-andy/Program.cs [hunk 4]
```diff
@@ -0,0 +1,592 @@
...
+            command.Parameters.AddWithValue("@units", units);
+            command.Parameters.AddWithValue("@quantity", quantity);
+
+            command.ExecuteNonQuery();
+        }
+        connection.Close();
+    }
+}
+
+string GetDateInput()
```
Comment (chrisjamiecarter) [inline HabitTracker.harris-andy/Program.cs:126]: There is some craziness going on here. Whilst the solution works fine. Why not return a `DateTime` type? This might be useful in future cases where you might need to validate dates etc.
I only say this because if I see a method called `GetDateInput()` I would expect a date return type. Maybe `GetInputDateAsString()` would be a better name.

File: HabitTracker.harris-andy/Program.cs [hunk 5]
```diff
@@ -0,0 +1,592 @@
...
+void Insert()
+{
+    Console.Clear();
+    string date = GetDateInput();
+    Console.Clear();
+    string hobby = GetHobby();
+    Console.Clear();
+    string units = GetUnitsInput();
+    Console.Clear();
+    int quantity = GetQuantityInput();
```
Comment (chrisjamiecarter) [inline HabitTracker.harris-andy/Program.cs:108]: 2 part comment. Part 1: At this point you are aware of the units. So you could pass them into `GetQuantityInput(units)`, see Part 2.

File: HabitTracker.harris-andy/Program.cs [hunk 6]
```diff
@@ -0,0 +1,592 @@
...
+        if (temp != null)
+        {
+            units = temp.Trim().ToLower();
+        }
+        if (units == "0") MainMenu();
+    }
+    return units;
+}
+
+int GetQuantityInput()
```
Comment (chrisjamiecarter) [inline HabitTracker.harris-andy/Program.cs:158]: 2 part comment. Part 2: If you accept the units `GetQuantityInput(string units)`, you could clear up your message to  `string message = $"Enter amount of {units} (greater than 0). Or press 0 to return to Main Menu";`

File: HabitTracker.harris-andy/Program.cs [hunk 7]
```diff
@@ -0,0 +1,592 @@
...
+            if (sortOption == 1) Console.WriteLine(outputByID);
+            if (sortOption == 2) Console.WriteLine(outputByDate);
+        }
+        Console.WriteLine("--------------------------------------------------\n");
+    }
+}
+
+void GetRecordSummary()
+{
+    (string searchTerm, string searchTermCategory) = GetSearchTerm();
```
Comment (chrisjamiecarter) [inline HabitTracker.harris-andy/Program.cs:235]: Instead of using a tuple here this is calling out for some OOP. A SeachTerm class or similar.

File: HabitTracker.harris-andy/HabitTracker.db
Comment (chrisjamiecarter) [inline HabitTracker.harris-andy/HabitTracker.db:1]: This file ideally should be added to gitignore. The actual database should not be part of source control. Also as the project requires to create the db if not exist and also you give option to populate with fake data.
Note: Just to clarify - The means/scripts/code to generate the database can be source control (and should be).

**Review by chrisjamiecarter — APPROVED** [review]
Hey @harris-andy 👋,

Apologies for reviewing another one of your projects 😢 ! I promise this will be the last 😀 .

Regardless, excellent work on another project submission 🎉 !

As always, I have performed a peer review and have some praise and suggestions:

- Project Directory Structure - Spot on 🏅 !
- README.md - Perfect content, well written ⭐ .
- Code Style - Improved massively 🚀 .
- Parameterised SQL Queries - Love it. No SQL injection attacks for you 🍨 !
- I love the relaxed and casual UI tone 🍺 .
- Challenged Requirements - Well done completing these 🥇 .
- Nice job on the Record Summary 🌟 .

**User Experience**

- You could clear the console each time before you write the menu. This would allow the user to segment an action, maintain readability of the console window and improve the experience for the user.
- If you present a list of information, like the habits, implement a `Console.WriteLine("Press any key to continue...");` and a `Console.Read();` after you have displayed the data. This will allow the user to read and digest the list before going back to a clean main menu (previous point).

Feel free to reach out if you have any questions or if you'd like to collaborate further on any of these points.

I will go ahead and mark as approved, keep up the excellent work on the next projects! 😊

Best regards,
@chrisjamiecarter

**Comment by harris-andy** [comment]
Hi Chris,

Thanks for the feedback! That's all very helpful! I got rather confused
about implementing OOP here. But after submitting the project I thought
about it more and came up with one option below:

class UserInput{
     MainMenu()
     GetDateInput()
     GetUnitsInput()
     GetQuantityInput()
     GetHobby()
     GetSearchTerm()

     GetRandomDate()??
     validateNumberEntry()??
}

class EditDatabase{
     Insert()
     Delete()
     Update()
     PopulateDatabase()
     DeleteTableContents()
}

class GetRecords {
     GetAllRecords()
     GetRecordSummary()
}

Each of these methods relies using the methods in the other classes, but I
couldn't figure out how to access them without creating a new instance. But
in a video on oop and software design I watched yesterday I saw that you
can pass one class into another to "connect" them. Would something like
this work?

class UserInput {
     MainMenu(GetRecords records, EditDatabase edit)
     GetDateInput(GetRecords records, EditDatabase edit)
     ...
}

class EditDatabase {
     Insert(GetRecords records, UserInput user)
     Delete(GetRecords records, UserInput user)
     ...
}

class GetRecords {
     GetAllRecords(EditDatabase edit, UserInput user)
     GetRecordSummary(EditDatabase edit, UserInput user)
}

I'd really like to get better at organizing these programs :) I've taken
note of your other feedback and will make sure to incorporate it into the
next project!

If it's easier to talk over discord please let me know and I'll message you
there.

Thanks again, I really appreciate the help!

Andy

**Comment by chrisjamiecarter** [comment]
Honestly, the best thing you can do is experiment. And see what works and what doesn't work. You need to make choices, then when you come back to refactor or add additional functionality you might get a lightbulb moment of "oh that's what I should have done!".

Take a look at my submission.
It is a lot of files, and may be hard to follow at first.
https://github.com/chrisjamiecarter/habit-tracker

But essentially I have a HabitTracker project, which is for domain stuff, or stuff that will be used everywhere and will always be the same.
Then I have a HabitTracker.Data project, which only handles database stuff, creating the database, adding, getting, updating, deleting records etc.
And finally I have a HabitTracker.ConsoleApp project, which is the User Interface. It displays views, gets user input, but importantly is the gateway to the Data project.

This means that if I wanted to put a web app UI on, I could re-use the HabitTracker and HabitTracker.Data projects.
If I wanted to change the Database provider from sqlite to SQL Server I only have to change the HabitTracker.Data project.

Hopefully this helps. Most important is that you have completed a project and you are on to the next one! 😄 

All the best
@chrisjamiecarter

**Comment by harris-andy** [comment]
I'm plodding away at the coding tracker now. Trying to implement the multiple classes and so far so good. I looked at your project and got immediately lost :) (and a little overwhelmed if I'm honest - that kind of complexity melts my brain at this stage) I see the value in your structure but following it is a little tricky.

Hopefully with more practice I'll get there. Thanks again!
